libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,249 @@
+	err := os.MkdirAll(settings.directoryPath(), os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't create disk queue directory: %w", err)
+	}
+
+	// Load the previous queue position, if any.
+	nextReadPosition, err := queuePositionFromPath(settings.stateFilePath())
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		// Errors reading / writing the position are non-fatal -- we just log a
+		// warning and fall back on the oldest existing segment, if any.
+		logger.Warnf("Couldn't load most recent queue position: %v", err)
+	}
+	positionFile, err := os.OpenFile(
+		settings.stateFilePath(), os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		// This is not the _worst_ error: we could try operating even without a
+		// position file. But it indicates a problem with the queue permissions on
+		// disk, which keeps us from tracking our position within the segment files
+		// and could also prevent us from creating new ones, so we treat this as a
+		// fatal error on startup rather than quietly providing degraded
+		// performance.
+		return nil, fmt.Errorf("Couldn't write to state file: %v", err)
+	}
+
+	// Index any existing data segments to be placed in segments.reading.
+	initialSegments, err := scanExistingSegments(settings.directoryPath())
+	if err != nil {
+		return nil, err
+	}
+	var nextSegmentID segmentID
+	if len(initialSegments) > 0 {
+		// Initialize nextSegmentID to the first ID after the existing segments.
+		lastID := initialSegments[len(initialSegments)-1].id
+		nextSegmentID = lastID + 1
+	}
+
+	// If any of the initial segments are older than the current queue
+	// position, move them directly to the acked list where they can be
+	// deleted.
+	ackedSegments := []*queueSegment{}
+	readSegmentID := nextReadPosition.segmentID
+	for len(initialSegments) > 0 && initialSegments[0].id < readSegmentID {
+		ackedSegments = append(ackedSegments, initialSegments[0])
+		initialSegments = initialSegments[1:]
+	}
+
+	// If the queue position is older than all existing segments, advance
+	// it to the beginning of the first one.
+	if len(initialSegments) > 0 && readSegmentID < initialSegments[0].id {
+		nextReadPosition = queuePosition{segmentID: initialSegments[0].id}
+	}
+
+	queue := &diskQueue{
+		logger:   logger,
+		settings: settings,
+
+		segments: &diskQueueSegments{
+			reading:        initialSegments,
+			nextID:         nextSegmentID,
+			nextReadOffset: nextReadPosition.offset,
+		},
+
+		acks: newDiskQueueACKs(logger, nextReadPosition, positionFile),
+
+		readerLoop:  newReaderLoop(settings),
+		writerLoop:  newWriterLoop(logger, settings),
+		deleterLoop: newDeleterLoop(settings),
+
+		producerWriteRequestChan: make(chan producerWriteRequest),
+
+		done: make(chan struct{}),
+	}
+
+	// We wait for four goroutines on shutdown: core loop, reader loop,
+	// writer loop, deleter loop.
+	queue.waitGroup.Add(4)
+
+	// Start the goroutines and return the queue!
+	go func() {
```

Comment:
These `go` function starts can be replaced by `unison.TaskGroup`. This way we could get rid of several lines of boilerplate.

```golang
queue.taskGroup.Go(func (canceller unison.Canceller) error {
    queue.readerLoop.run(canceller)
})

queue.taskGroup.Go(func (canceller unison.Canceller) error {
    queue.writerLoop.run(canceller)
})

queue.taskGroup.Go(func (canceller unison.Canceller) error {
    queue.deleterLoop.run(canceller)
})

queue.taskGroup.Go(func (canceller unison.Canceller) error {
    queue.run(canceller)
})
```

Also, in the `Close` function we could simply call `dq.taskGroup.Stop` to stop and wait for all functions to finish. We could also get rid of the channel, as `unison.TaskGroup` cancels all goroutines for us.
```golang
func (dq *diskQueue) Close() error {
	dq.taskGroup.Stop()

	return nil
}
```

Furthermore, instead of using done channels we could use contexts. So

```golang
for {
    select {
        case <-done:
            return
       default:
            /// do your thing
    }
}
```
can become this:

```golang
for cancelCtx.Err() == nil {
    // do you thing
}
```

See more: https://github.com/elastic/go-concert/blob/master/unison/taskgroup.go#L72

Also, these patterns are already part of Beats, we would not be introducing any new practices here. Just replacing old, inferior stuff.

Reply:
I don't think that library works well in this case -- a task group shares a single closer, which propagates the signal to all the goroutines, but that isn't what we want. We have to be careful about the order the helpers are shut down and that we clean up as we go (see the ordering in `core_loop.go:diskQueue.handleShutdown`), so we don't want something that manages them for us, we need to signal the core loop first and do the rest manually.

(Beyond that, it would also add an extra field to all the helper loops and an extra case to many select statements, without giving any more functionality since most of the time we will need to use the canceler as a channel anyway -- the `select { <-done: default: ... }` pattern is annoying but it's also the less common case, and as things stand right now most of the code doesn't need to keep track of an extra done channel / canceler, which I think is a good tradeoff against the boilerplate of a few manual `waitGroup.Done()` calls.)